docs/proposals/pod-troubleshooting.md

```diff
@@ -0,0 +1,489 @@
+
+The disadvantage of using containers built `FROM scratch` is the lack of system
+binaries provided a base Operation System image makes it difficult to
+troubleshoot running containers. Kubernetes should enable troubleshooting pods
+regardless of the contents of the container image.
+
+## Goals and Non-Goals
+
+Goals include:
+
+*   Enable troubleshooting of minimal container images
+*   Allow troubleshooting of containers in `CrashLoopBackoff`
+*   Improve supportability of native Kubernetes applications
+*   Enable novel uses of short-lived containers in pods
+
+Non-Goals of this proposal are:
+
+*   Fully design the ability to cross-mount container image filesystems. This is
+    complex and should be covered in a separate doc.
+*   Guarantee resources for ad-hoc troubleshooting. If troubleshooting causes a
+    pod to exceed its resource limit it may be evicted.
+
+## Requirements
+
+A solution to troubleshoot minimal container images MUST:
+
+*   troubleshoot arbitrary running containers with minimal prior configuration
+*   allow access to all namespaces shared by containers in a pod and the mount
+    namespaces of individual containers
+*   fetch troubleshooting utilities at debug time rather than at the time of pod
+    initialization
+*   respect admission restrictions
+*   allow introspection of pod state using existing tools (no hidden containers)
+*   support arbitrary runtimes via the CRI (possibly with reduced feature set)
+
+A good solution SHOULD:
+
+*   have an excellent user experience (i.e. should be a feature of the platform
+    rather than a config-time solution)
+*   require no direct access to the node
+*   have no inherent side effects to the running container image
+*   re-use existing container image distribution channels
+*   enable detection of pods that have been modified by an exec ("tainted" pods)
+
+## kubectl debug
+
+We will introduce an "execute a debug container in a pod" pattern that parallels
+`kubectl exec`'s "execute a debug process in a container" in a new command,
+`kubectl debug`, which has two modes of operation:
+
+1.  *Copy Debug* mode creates a copy of a pod with minor changes to the pod spec
+    specified on the command line.
+1.  *Running Debug* mode causes Kubernetes to run a new *debug container* in the
+    pod context. A *debug container* is not part of the pod spec and cannot be
+    configured by the user. It is created by a debug operation on an existing
+    pod and reported in `PodStatus`.
+
+The mode distinction is for the purposes of this proposal, and the end user need
+not be aware of them. They form a single debugging command that allows the user
+to modify a running pod if possible and create a copy otherwise. The
+determination is made based on container name, for example:
+
+```
+% kubectl run example --image=nginx --restart=Never
+pod "example" created
+
+# Container "shell" does not exist, add a new container to running pod
+kubectl debug -p example -c shell --image=debian
```

Comment:
Specifying a new pod name to debug is an edge case.

All debugging is about the target, so the argument should be a single value which supports our consistent "acts on pod-like things" - deployments, RS, or pods (with pod being the default resource type).  There will have to be controller specific behavior for many cases, but that's actually desirable (we want stateful set debug not to crash the set).  That also leaves the door open to future semantic options - "--under-service=foo", for example.

A default image is probably not practical for many reasons, starting with library compatibility all the way up to security concerns.  

We might be trying too hard to cram "create a copy to debug" and "add a container" under the same verb.  "Add a container" feels more like an attach variant.